web-app/src/main/java/com/google/sps/data/QuadTree.java

```diff
@@ -59,4 +58,76 @@ public void printTree() {
     }
     System.out.printf("%n");
   }
+
+  public void insert(PoliceReport report) {
+    Node currentNode = root;
+    double reportLat = report.getLat();
+    double reportLng = report.getLng();
+
+    // Find a leaf to insert the report in
+    while (!currentNode.leaf) {
+      currentNode.numReports += 1;
+      for (Node child : currentNode.children) {
+        if (child.coordinates.inRectangle(reportLat, reportLng)) {
+          currentNode = child;
+          break;
+        }
+      }
+    }
+
+    currentNode.numReports += 1;
+    currentNode.reports.add(report);
+    // If max capacity has been exceeded, create child nodes
+    if (currentNode.numReports > reportCapacity && currentNode.depth < maxDepth) {
+      currentNode.children = reallocateReports(currentNode.reports, currentNode.coordinates, currentNode.depth);
+      currentNode.leaf = false;
+      currentNode.reports = null;
+    }
+  }
+
+  private Node[] reallocateReports(ArrayList<PoliceReport> reports, Rectangle coordinates, int depth) {
+    Node[] children = new Node[4];
+
+    for (Direction direction : Direction.values()) {
+      Rectangle newCoordinates;
+      ArrayList<PoliceReport> newReports = new ArrayList<PoliceReport>();
+      int newDepth = depth + 1;
+
+      switch (direction) {
+        case NW:
+          newCoordinates = coordinates.getNW();
+          break;
+        case NE:
+          newCoordinates = coordinates.getNE();
+          break;
+        case SE:
+          newCoordinates = coordinates.getSE();
+          break;
+        case SW:
+          newCoordinates = coordinates.getSW();
+          break;
+        default:
+          System.out.println("Unexpected case in switch statement");
+          return children;
+      }
+
+      for (PoliceReport report : reports) {
+        if (newCoordinates.inRectangle(report.getLat(), report.getLng())) {
```

Comment:
is there the possibility that a report gets added to multiple children since inRectangle may return true if it's on the edge?

Instead, can you create the 4 empty children first, then loop through each report and put it in the right child? That'll guarantee each report is only put into exactly one child.